src/core/UniswapV4.sol

```diff
@@ -295,10 +301,34 @@ abstract contract UniswapV4 is SettlerAbstract {
             unchecked {
                 params.amountSpecified = int256((state.sell().amount() * bps).unsafeDiv(BASIS)).unsafeNeg();
             }
-            // TODO: price limits
-            params.sqrtPriceLimitX96 = uint160(
-                (!zeroForOne).ternary(uint160(1461446703485210103287273052203988822378723970341), uint160(4295128740))
-            );
+            
+            {
+                // priceSqrtX96 is uint160
+                // uint256 used in favor of future operations that will overflow uint160
+                uint256 priceSqrtX96 = IPoolManager(msg.sender).unsafeSqrtPriceX96(key);
+                // Factor is:
+                // 1. 28011385487393069959365969113 approximately floor(sqrt(2**188)) (95 bits)
+                //    which is equivalent to 1 / sqrt(2) in Q65.95
+                //    2**95 / sqrt(2) = sqrt(2) * (2**94) = sqrt(2**188)
+                //    therefore it is the largest integer that multiplied by itself doesn't exceed 2**188
+                // 2. 56022770974786139918731938227 approximately floor(sqrt(2**191)) (96 bits)
+                //    which is equivalent to sqrt(2) in Q65.95
+                //    sqrt(2) * (2**95) = sqrt(2**191)
+                //    therefore it is the largest integer that multiplied by itself doesn't exceed 2**191
+                // Q65.95 was used instead of Q64.96 to prevent uint256 overflows later on as sqrt(2) in Q64.96 would be 97 bits
+                uint256 factor = zeroForOne.ternary(uint256(28011385487393069959365969113), uint256(56022770974786139918731938227));
+
+                unchecked {
+                    // no overflow when multiplying as factors are 160 bits and at most 96 bits respectively
+                    // shifted right 95 bitsto keep the price as Q64.96
+                    priceSqrtX96 = (priceSqrtX96 * factor) >> 95;
+                }
+                
+                uint256 limit = (!zeroForOne).ternary(uint256(1461446703485210103287273052203988822378723970341), uint256(4295128740));
+                (uint256 lo, uint256 hi) = zeroForOne.maybeSwap(priceSqrtX96, limit);
+                // All operations where rounded down to ensure that the selected price has at most 100% price impact
```

Comment:
I'm not sure this comment is justified. Can you explain more?

Reply:
🤔, it is kind of naive, the math was done to get to the price which exact 100% price impact, as this is exact math, so in floating math we will get to the exact price (`==`). But, as this is exact math, rounding down in the middle changes the inequality to less or equal (`<=`). That is why I said `at most` but we should not be able to reach the boundary

Reply:
> to the price which exact 100% price impact

are you sure it's _exactly_ 100% price impact? you're using fixnums, which are not typically exact and demand attention to rounding direction. given that the direction that the price moves depends on `zeroForOne`, I would expect the correct computation to require that we round up in one direction and round down in the other. I would not expect that rounding down gives the correct `<=` bounds in both cases.

of course, given that we're working with fairly large fixnum bases here (Q96 and Q95), it's possible that the error introduced by inconsistent rounding either doesn't _practically_ matter or literally does not matter even at the extremes of  `sqrtPriceX96`. it would be good to call this out explicitly in this comment.